File: problems/problems_3579/solution.py
from collections import defaultdict
from math import inf

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.minOperations(*test_input)

    def minOperations(self, word1: str, word2: str) -> int:
        def update(counter, x, y):
            if x == y:
                return
            if counter[(y, x)] > 0:
                counter[(y, x)] -= 1
                # 因为刚刚有取到逆对, 所以用交换操作代替原来的替换操作, 操作数不需要变化了
            else:
                counter[(x, y)] += 1
                nonlocal op
                op += 1 # 暂时使用替换操作

        n = len(word1)

        # 预处理所有反转操作子串
        rev_op = [[0] * n for _ in range(n)]
        # 中心扩展法
        for i in range(2 * n - 1):
            cnt = defaultdict(int)
            op = 1 # 反转操作
            l, r = i // 2, (i+1) // 2
            while l >= 0 and r < n:
                update(cnt, word1[l], word2[r])
                if l != r:
                    update(cnt, word1[r], word2[l])
                rev_op[l][r] = op
                l -= 1
                r += 1

        f = [0] * (n + 1)
        for i in range(n):
            res = inf
            cnt = defaultdict(int)
            op = 0 # 正序不操作
            for j in range(i, -1, -1):
                update(cnt, word1[j], word2[j])
                res = min(res, f[j] + min(op, rev_op[j][i])) # 子串[j, i]的最小操作数
            f[i + 1] = res
        return f[n]

File: problems/problems_3579/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

#include <unordered_map>


using namespace std;
using json = nlohmann::json;

class Solution {
public:
    int minOperations(string word1, string word2) {
        auto update = [](unordered_map<int, int>& cnt, char a, char b, int& op) {
            if (a == b) {
                return;
            }
            int h = (b - 'a') * 26 + (a - 'a');
            if (cnt[h] > 0) {
                cnt[h]--;
            } else {
                cnt[(a-'a') * 26 + (b - 'a')]++;
                ++op;
            }
        };
        int n = word1.length();

        vector<vector<int>> rev_op(n, vector<int>(n, 0));
        for (int i = 0; i < 2 * n - 1; ++i) {
            int l = i / 2, r = (i + 1) / 2;
            int op = 1; // reverse operation
            unordered_map<int, int> cnt;
            while (l >= 0 && r < n) {
                update(cnt, word1[l], word2[r], op);
                if (l != r) {
                    update(cnt, word1[r], word2[l], op);
                }
                rev_op[l][r] = op;
                --l;
                ++r;
            }
        }

        vector<int> dp(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            int op = 0;
            int res = INT_MAX;
            unordered_map<int, int> cnt;
            for (int j = i; j >= 0; --j) {
                update(cnt, word1[j], word2[j], op);
                res = min(res, dp[j] + min(op, rev_op[j][i]));
            }
            dp[i + 1] = res;
        }
        return dp[n];
    }
};

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	Solution solution;
	string word1 = json::parse(inputArray.at(0));
	string word2 = json::parse(inputArray.at(1));
	return solution.minOperations(word1, word2);
}
